### PR TITLE
User-friendly (ish) interface for metadata extraction

### PR DESCRIPTION
**Pull request by richfitz**
This has been requested in a few guises over the years, and I think it will perhaps form the basis of other tooling too.

Extract some fields from the metadata of a bunch of packets. The implementation is fairly rich, and is described in some detail in the docs (which I used to sketch out the interface before implementation).  I think this strikes a reasonable balance of usability, generalisability and extensibility, but very open to suggestions.

The query syntax is deliberately not-very R-ish as we might want to adopt this into the rust tool in some guise?

This is a big PR, but quite a lot of it is docs, error handling and tests. The docs (in the Rd file) is probably the best place to start

### CODE REVIEW

**Review by r-ash — CHANGES_REQUESTED** [review]
Some small suggested changes but looks good.

Guess big question I'm interested in is why you wanted the type conversion to be explicit instead of leaving it up to R to convert (which generally works fairly well). Just thinking ahead to other language implementations of this where the automatic conversion might not be possible or the same?

**Comment by richfitz** [comment]
> Guess big question I'm interested in is why you wanted the type conversion to be explicit instead of leaving it up to R to convert (which generally works fairly well). Just thinking ahead to other language implementations of this where the automatic conversion might not be possible or the same?

I don't think this works well in practice (thinking here of why we avoid `sapply` in general). In general very little of this is knowable and we'll need the user to guide us. Trying to be chill about it will mean that different calls will result in different classes (for example, pulling `parameters` might be a character vector in one case because there's only one type of parameter and its character, a list in others and a list of lists in another). This way, the same call will be type stable even when the underlying  data changes

File: R/outpack_tools.R [hunk 1]
```diff
@@ -0,0 +1,356 @@
...
+##' extract = c("name", "parameters", "files")
+##' ```
+##'
+##' This creates a data.frame with columns corresponding to these
+##' keys, one row per packet. Because `name` is always a string, it
+##' will be a character vector, but because `parameters` and `files`
+##' are more complex, these will be list columns.
+##'
+##' You must not provide `id`; it is always returned and always first
+##' as a character vector column.
```
[the remark this answers is not in the document]
Comment (richfitz) [reply R/outpack_tools.R:91]: currently, yes it errored. I wondered about being nice and just skipping over it but decided that every time I try and be helpful it creates complicated things to maintain later, so just threw instead. We can always walk that back later

File: R/outpack_tools.R [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+  extract <- parse_extract(extract, environment())
+
+  meta <- lapply(ids, root$metadata, full = TRUE)
+
+  env <- environment()
+  ret <- data_frame(id = ids)
+  for (i in seq_len(nrow(extract))) {
+    from_i <- extract$from[[i]]
+    is_i <- extract$is[[i]]
+    value_i <- lapply(meta, function(x) x[[from_i]])
```
[the remark this answers is not in the document]
Comment (richfitz) [reply R/outpack_tools.R]: this did remind me though that I wanted to prevent the opaque error you get when this fails

**Review by r-ash — APPROVED** [review]

**Pull request merged**